<a href="https://colab.research.google.com/github/BabimBeam/BSC_DPDM2025/blob/main/Midterm_663020581_4_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

[บันทึกการใช้ AI](https://docs.google.com/document/d/15P6coh1PgYrjGhRvZhzmegMf7QSj9a35DbYc2_epc7I/edit?usp=sharing)


[Output](https://docs.google.com/spreadsheets/d/1mfsdgWVcli2MbMupTJw68XwFVqxk6W5oM-F4-iS9R8s/edit?usp=sharing)

# 1.Import Data  นำเข้าข้อมูล

In [1]:
import pandas as pd

In [2]:
import os

เชื่อมต่อ Google drive

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
hii_daily_rain = pd.read_csv('/content/drive/MyDrive/BSC_DPDM/BSC DPDM midterm 2025/Daily rain - (Statistics 2012-2024)/hii_daily_rain.csv')
hii_daily_rain.head()

,station_code,measure_datetime,data,station_name,latitude,longitude,quality_flag,basin,sub_basin,tambon,amphoe,province,data_type
0,ABRT,2012-01-01 07:00:00+07,0.0,วัดเวฬุวัน,16.054785,103.66429,U,ชี,ลำน้ำชีส่วนที่ 4/2,เหนือเมือง,เมืองร้อยเอ็ด,ร้อยเอ็ด,rainfall_daily
1,ABRT,2012-01-03 07:00:00+07,0.0,วัดเวฬุวัน,16.054785,103.66429,U,ชี,ลำน้ำชีส่วนที่ 4/2,เหนือเมือง,เมืองร้อยเอ็ด,ร้อยเอ็ด,rainfall_daily
2,ABRT,2012-01-06 07:00:00+07,0.0,วัดเวฬุวัน,16.054785,103.66429,U,ชี,ลำน้ำชีส่วนที่ 4/2,เหนือเมือง,เมืองร้อยเอ็ด,ร้อยเอ็ด,rainfall_daily
3,ABRT,2012-01-07 07:00:00+07,0.0,วัดเวฬุวัน,16.054785,103.66429,U,ชี,ลำน้ำชีส่วนที่ 4/2,เหนือเมือง,เมืองร้อยเอ็ด,ร้อยเอ็ด,rainfall_daily
4,ABRT,2012-01-08 07:00:00+07,0.0,วัดเวฬุวัน,16.054785,103.66429,U,ชี,ลำน้ำชีส่วนที่ 4/2,เหนือเมือง,เมืองร้อยเอ็ด,ร้อยเอ็ด,rainfall_daily


## คำอธิบายคอลัมน์ของ DataFrame `hii_daily_rain`

*   `station_code`: รหัสเฉพาะสำหรับสถานีตรวจวัดสภาพอากาศ
*   `measure_datetime`: วันที่และเวลาที่ทำการตรวจวัดปริมาณน้ำฝน
*   `data`: ปริมาณน้ำฝนที่ตรวจวัดได้รายวัน
*   `station_name`: ชื่อสถานีตรวจวัดสภาพอากาศ
*   `latitude`: ละติจูดทางภูมิศาสตร์ของสถานีตรวจวัดสภาพอากาศ
*   `longitude`: ลองจิจูดทางภูมิศาสตร์ของสถานีตรวจวัดสภาพอากาศ
*   `quality_flag`: ธงที่ระบุคุณภาพของข้อมูล (เช่น 'U' สำหรับข้อมูลที่ยังไม่ได้รับการตรวจสอบหรือยืนยัน)
*   `basin`: ลุ่มน้ำหลักที่สถานีตั้งอยู่
*   `sub_basin`: ลุ่มน้ำย่อยภายในลุ่มน้ำหลัก
*   `tambon`: หน่วยการปกครองย่อย 'ตำบล'
*   `amphoe`: หน่วยการปกครอง 'อำเภอ'
*   `province`: หน่วยการปกครอง 'จังหวัด'
*   `data_type`: ประเภทของข้อมูล ซึ่งในกรณีนี้คือ 'rainfall_daily' (ปริมาณน้ำฝนรายวัน)

In [24]:
Stations_model_HII = pd.read_csv('/content/drive/MyDrive/BSC_DPDM/BSC DPDM midterm 2025/Input - (Operation Daily rain from API_ 2025)/Stations in the model_HII.csv')
Stations_model_HII.head()

,Node,Lat,Long
0,ACRU,15.788054,104.642235
1,BBHN,9.536913,98.579490
2,BBUA,17.372976,103.984436
3,BCNG,17.386840,103.290870
4,BDAR,14.587073,102.495190


## คำอธิบายคอลัมน์ของ DataFrame `Stations_model_HII`

*   `Node`: รหัสหรือชื่อสถานีตรวจวัด
*   `Lat`: ละติจูดทางภูมิศาสตร์ของสถานี
*   `Long`: ลองจิจูดทางภูมิศาสตร์ของสถานี

In [6]:
data_folder_path = '/content/drive/MyDrive/BSC_DPDM/BSC DPDM midterm 2025/Input - (Operation Daily rain from API_ 2025)'
folder_contents = os.listdir(data_folder_path)
print(f"Contents of the data folder: {folder_contents}")

Contents of the data folder: ['Stations in the model_HII.csv', '2025-08', '2025-10', '2025-07', '2025-09', '2025-12', '2025-11']


In [7]:
import os

folder_path_2025_07 = '/content/drive/MyDrive/BSC_DPDM/BSC DPDM midterm 2025/Input - (Operation Daily rain from API_ 2025)/2025-07'
contents_2025_07 = os.listdir(folder_path_2025_07)
print(f"Contents of {folder_path_2025_07}:\n{contents_2025_07}")

Contents of /content/drive/MyDrive/BSC_DPDM/BSC DPDM midterm 2025/Input - (Operation Daily rain from API_ 2025)/2025-07:
['HII_daily_rainfall_2025-07-09.csv', 'HII_daily_rainfall_2025-07-10.csv', 'HII_daily_rainfall_2025-07-02.csv', 'HII_daily_rainfall_2025-07-01.csv', 'HII_daily_rainfall_2025-07-06.csv', 'HII_daily_rainfall_2025-07-11.csv', 'HII_daily_rainfall_2025-07-03.csv', 'HII_daily_rainfall_2025-07-08.csv', 'HII_daily_rainfall_2025-07-07.csv', 'HII_daily_rainfall_2025-07-04.csv', 'HII_daily_rainfall_2025-07-05.csv', 'HII_daily_rainfall_2025-07-14.csv', 'HII_daily_rainfall_2025-07-25.csv', 'HII_daily_rainfall_2025-07-24.csv', 'HII_daily_rainfall_2025-07-13.csv', 'HII_daily_rainfall_2025-07-20.csv', 'HII_daily_rainfall_2025-07-27.csv', 'HII_daily_rainfall_2025-07-18.csv', 'HII_daily_rainfall_2025-07-26.csv', 'HII_daily_rainfall_2025-07-17.csv', 'HII_daily_rainfall_2025-07-30.csv', 'HII_daily_rainfall_2025-07-29.csv', 'HII_daily_rainfall_2025-07-31.csv', 'HII_daily_rainfall_2025-07

In [8]:
import pandas as pd
import os

folder_path_2025_07 = '/content/drive/MyDrive/BSC_DPDM/BSC DPDM midterm 2025/Input - (Operation Daily rain from API_ 2025)/2025-07'
contents_2025_07 = os.listdir(folder_path_2025_07)

# Filter for CSV files and sort them to ensure correct order if needed
csv_files_2025_07 = sorted([f for f in contents_2025_07 if f.endswith('.csv')])

# List to hold individual dataframes
df_list_2025_07 = []

# Read each CSV file and append to the list
for file_name in csv_files_2025_07:
    file_path = os.path.join(folder_path_2025_07, file_name)
    df = pd.read_csv(file_path)
    df_list_2025_07.append(df)

# Concatenate all dataframes into one
API07 = pd.concat(df_list_2025_07, ignore_index=True)

print(f"Successfully loaded {len(df_list_2025_07)} CSV files into a single DataFrame 'API07'.")
print("First 5 rows of API07:")
print(API07.head())

Successfully loaded 31 CSV files into a single DataFrame 'API07'.
First 5 rows of API07:
           rainfall_datetime     rainfall_datetime_calc  rainfall_value  \
0  2025-07-02 00:00:00+07:00  2025-07-02T07:00:00+07:00             0.0   
1  2025-07-02 00:00:00+07:00  2025-07-02T07:00:00+07:00             0.0   
2  2025-07-02 00:00:00+07:00  2025-07-02T07:00:00+07:00            34.0   
3  2025-07-02 00:00:00+07:00  2025-07-02T07:00:00+07:00             0.2   
4  2025-07-02 00:00:00+07:00  2025-07-02T07:00:00+07:00            38.2   

   tele_station_id tele_station_oldcode measure_datetime  
0              589                 MWOG       2025-07-01  
1              564                 MSRM       2025-07-01  
2                4               CPY015       2025-07-01  
3              867                 KMSI       2025-07-01  
4               54               CPY016       2025-07-01  


In [9]:
import pandas as pd
import os

folder_path_2025_08 = os.path.join(data_folder_path, '2025-08')
contents_2025_08 = os.listdir(folder_path_2025_08)

# Filter for CSV files and sort them to ensure correct order if needed
csv_files_2025_08 = sorted([f for f in contents_2025_08 if f.endswith('.csv')])

# List to hold individual dataframes
df_list_2025_08 = []

# Read each CSV file and append to the list
for file_name in csv_files_2025_08:
    file_path = os.path.join(folder_path_2025_08, file_name)
    df = pd.read_csv(file_path)
    df_list_2025_08.append(df)

# Concatenate all dataframes into one
API08 = pd.concat(df_list_2025_08, ignore_index=True)

print(f"Successfully loaded {len(df_list_2025_08)} CSV files into a single DataFrame 'API08'.")
print("First 5 rows of API08:")
print(API08.head())

Successfully loaded 32 CSV files into a single DataFrame 'API08'.
First 5 rows of API08:
           rainfall_datetime     rainfall_datetime_calc  rainfall_value  \
0  2025-08-02 00:00:00+07:00  2025-08-02T07:00:00+07:00             0.0   
1  2025-08-02 00:00:00+07:00  2025-08-02T07:00:00+07:00             0.0   
2  2025-08-02 00:00:00+07:00  2025-08-02T07:00:00+07:00             0.2   
3  2025-08-02 00:00:00+07:00  2025-08-02T07:00:00+07:00             0.0   
4  2025-08-02 00:00:00+07:00  2025-08-02T07:00:00+07:00             0.0   

   tele_station_id tele_station_oldcode measure_datetime station_code  value  
0            493.0               YOM004       2025-08-01          NaN    NaN  
1            495.0               VLGE36       2025-08-01          NaN    NaN  
2            496.0                 SNGT       2025-08-01          NaN    NaN  
3            497.0                 WCHN       2025-08-01          NaN    NaN  
4            498.0                 NMKI       2025-08-01         

In [10]:
import pandas as pd
import os

folder_path_2025_09 = os.path.join(data_folder_path, '2025-09')
contents_2025_09 = os.listdir(folder_path_2025_09)

# Filter for CSV files and sort them to ensure correct order if needed
csv_files_2025_09 = sorted([f for f in contents_2025_09 if f.endswith('.csv')])

# List to hold individual dataframes
df_list_2025_09 = []

# Read each CSV file and append to the list
for file_name in csv_files_2025_09:
    file_path = os.path.join(folder_path_2025_09, file_name)
    df = pd.read_csv(file_path)
    df_list_2025_09.append(df)

# Concatenate all dataframes into one
API09 = pd.concat(df_list_2025_09, ignore_index=True)

print(f"Successfully loaded {len(df_list_2025_09)} CSV files into a single DataFrame 'API09'.")
print("First 5 rows of API09:")
print(API09.head())

Successfully loaded 32 CSV files into a single DataFrame 'API09'.
First 5 rows of API09:
           rainfall_datetime     rainfall_datetime_calc  rainfall_value  \
0  2025-09-02 00:00:00+07:00  2025-09-02T07:00:00+07:00            15.6   
1  2025-09-02 00:00:00+07:00  2025-09-02T07:00:00+07:00             0.4   
2  2025-09-02 00:00:00+07:00  2025-09-02T07:00:00+07:00             0.2   
3  2025-09-02 00:00:00+07:00  2025-09-02T07:00:00+07:00             0.6   
4  2025-09-02 00:00:00+07:00  2025-09-02T07:00:00+07:00             2.6   

   tele_station_id tele_station_oldcode measure_datetime station_code  value  
0              NaN                 ACRU       2025-09-01          NaN    NaN  
1              NaN               AIT001       2025-09-01          NaN    NaN  
2              NaN               AIT002       2025-09-01          NaN    NaN  
3              NaN               AIT005       2025-09-01          NaN    NaN  
4              NaN               AIT008       2025-09-01         

In [11]:
import pandas as pd
import os

folder_path_2025_10 = os.path.join(data_folder_path, '2025-10')
contents_2025_10 = os.listdir(folder_path_2025_10)

# Filter for CSV files and sort them to ensure correct order if needed
csv_files_2025_10 = sorted([f for f in contents_2025_10 if f.endswith('.csv')])

# List to hold individual dataframes
df_list_2025_10 = []

# Read each CSV file and append to the list
for file_name in csv_files_2025_10:
    file_path = os.path.join(folder_path_2025_10, file_name)
    df = pd.read_csv(file_path)
    df_list_2025_10.append(df)

# Concatenate all dataframes into one
API10 = pd.concat(df_list_2025_10, ignore_index=True)

print(f"Successfully loaded {len(df_list_2025_10)} CSV files into a single DataFrame 'API10'.")
print("First 5 rows of API10:")
print(API10.head())

Successfully loaded 31 CSV files into a single DataFrame 'API10'.
First 5 rows of API10:
           rainfall_datetime     rainfall_datetime_calc  rainfall_value  \
0  2025-10-02 00:00:00+07:00  2025-10-01T07:00:00+07:00             0.2   
1  2025-10-02 00:00:00+07:00  2025-10-01T07:00:00+07:00            10.8   
2  2025-10-02 00:00:00+07:00  2025-10-01T07:00:00+07:00             0.0   
3  2025-10-02 00:00:00+07:00  2025-10-01T07:00:00+07:00             2.4   
4  2025-10-02 00:00:00+07:00  2025-10-01T07:00:00+07:00             0.0   

   tele_station_id tele_station_oldcode measure_datetime  
0              812               TPI004       2025-10-01  
1              688                 TPKO       2025-10-01  
2              746               THA011       2025-10-01  
3          1198626               TRD005       2025-10-01  
4              227                 UDMS       2025-10-01  


In [12]:
import pandas as pd
import os

folder_path_2025_11 = os.path.join(data_folder_path, '2025-11')
contents_2025_11 = os.listdir(folder_path_2025_11)

# Filter for CSV files and sort them to ensure correct order if needed
csv_files_2025_11 = sorted([f for f in contents_2025_11 if f.endswith('.csv')])

# List to hold individual dataframes
df_list_2025_11 = []

# Read each CSV file and append to the list
for file_name in csv_files_2025_11:
    file_path = os.path.join(folder_path_2025_11, file_name)
    df = pd.read_csv(file_path)
    df_list_2025_11.append(df)

# Concatenate all dataframes into one
API11 = pd.concat(df_list_2025_11, ignore_index=True)

print(f"Successfully loaded {len(df_list_2025_11)} CSV files into a single DataFrame 'API11'.")
print("First 5 rows of API11:")
print(API11.head())

Successfully loaded 30 CSV files into a single DataFrame 'API11'.
First 5 rows of API11:
           rainfall_datetime     rainfall_datetime_calc  rainfall_value  \
0  2025-11-02 00:00:00+07:00  2025-11-01T07:00:00+07:00             1.6   
1  2025-11-02 00:00:00+07:00  2025-11-01T07:00:00+07:00            72.0   
2  2025-11-02 00:00:00+07:00  2025-11-01T07:00:00+07:00             0.0   
3  2025-11-02 00:00:00+07:00  2025-11-01T07:00:00+07:00            41.6   
4  2025-11-02 00:00:00+07:00  2025-11-01T07:00:00+07:00            61.2   

   tele_station_id tele_station_oldcode measure_datetime  
0              245               MUN009       2025-11-01  
1            12477               PRN002       2025-11-01  
2              850               JNA002       2025-11-01  
3              782                 MPBR       2025-11-01  
4           575572               BKC008       2025-11-01  


In [13]:
import pandas as pd
import os

folder_path_2025_12 = os.path.join(data_folder_path, '2025-12')
contents_2025_12 = os.listdir(folder_path_2025_12)

# Filter for CSV files and sort them to ensure correct order if needed
csv_files_2025_12 = sorted([f for f in contents_2025_12 if f.endswith('.csv')])

# List to hold individual dataframes
df_list_2025_12 = []

# Read each CSV file and append to the list
for file_name in csv_files_2025_12:
    file_path = os.path.join(folder_path_2025_12, file_name)
    df = pd.read_csv(file_path)
    df_list_2025_12.append(df)

# Concatenate all dataframes into one
API12 = pd.concat(df_list_2025_12, ignore_index=True)

print(f"Successfully loaded {len(df_list_2025_12)} CSV files into a single DataFrame 'API12'.")
print("First 5 rows of API12:")
print(API12.head())

Successfully loaded 31 CSV files into a single DataFrame 'API12'.
First 5 rows of API12:
           rainfall_datetime     rainfall_datetime_calc  rainfall_value  \
0  2025-12-02 00:00:00+07:00  2025-12-01T07:00:00+07:00             0.0   
1  2025-12-02 00:00:00+07:00  2025-12-01T07:00:00+07:00             0.0   
2  2025-12-02 00:00:00+07:00  2025-12-01T07:00:00+07:00             0.0   
3  2025-12-02 00:00:00+07:00  2025-12-01T07:00:00+07:00             0.0   
4  2025-12-02 00:00:00+07:00  2025-12-01T07:00:00+07:00             0.0   

   tele_station_id tele_station_oldcode measure_datetime  
0              475               NAN014       2025-12-01  
1              330               CHI005       2025-12-01  
2              333               CHI006       2025-12-01  
3              334                 KUDN       2025-12-01  
4              584               CPY002       2025-12-01  


รวมไฟล์ข้อมูลรายวันหลายๆ ไฟล์ ให้กลายเป็นตารางเดียว" (Data Aggregation) โดยเป็นการจัดการข้อมูลในระดับโฟลเดอร์

1. การเตรียมรายชื่อไฟล์
os.listdir: เข้าไปสแกนในโฟลเดอร์ 2025-07 ว่ามีไฟล์ชื่ออะไรบ้าง

List Comprehension: กรองเอาเฉพาะไฟล์ที่นามสกุล .csv เท่านั้น (กันเหนียวเผื่อมีไฟล์ประเภทอื่นปนอยู่)

sorted: เรียงลำดับชื่อไฟล์ (เช่น วันที่ 01, 02, 03...) เพื่อให้ข้อมูลเรียงต่อกันตามเวลาอย่างถูกต้อง

2. การอ่านข้อมูลเข้าสู่หน่วยความจำ
for loop: วนลูปเปิดไฟล์ CSV ทีละไฟล์ตามรายชื่อที่เตรียมไว้

os.path.join: สร้างที่อยู่ไฟล์ที่สมบูรณ์ (เช่น /.../2025-07/rain_01.csv)

df_list_2025_07.append(df): นำตารางของแต่ละวันไป "ฝากไว้ในตะกร้า" (List) ก่อน

3. การรวมร่าง (Concatenation)
pd.concat: นำตารางทั้งหมดในตะกร้ามา "วางต่อท้ายกัน" (Stacking) จนกลายเป็นตารางยาวตารางเดียวที่ชื่อว่า API07

ignore_index=True: สั่งให้รันเลขลำดับแถวใหม่ (0, 1, 2...) ต่อเนื่องกันไปเลย ไม่เอาเลข Index เก่าของแต่ละไฟล์มาปนกัน

ทำแบบนี้ไปเรื่อยๆจนถึงเดือนที่ 12

## Summary:

### Data Analysis Key Findings
*   Data for August 2025 (`API08`) was successfully compiled by concatenating 32 CSV files.
*   Data for September 2025 (`API09`) was successfully compiled by concatenating 32 CSV files.
*   Data for October 2025 (`API10`) was successfully compiled by concatenating 31 CSV files.
*   Data for November 2025 (`API11`) was successfully compiled by concatenating 30 CSV files.
*   Data for December 2025 (`API12`) was successfully compiled by concatenating 31 CSV files.


## คำอธิบายคอลัมน์ของ DataFrame `API`

*   `rainfall_datetime`: วันที่และเวลาที่มีการบันทึกปริมาณน้ำฝน (อาจเป็นเวลาเริ่มต้นของช่วงการวัด)
*   `rainfall_datetime_calc`: วันที่และเวลาที่ใช้ในการคำนวณปริมาณน้ำฝน (มักจะเป็นเวลา 07:00:00+07:00 ของวันที่เกิดฝน)
*   `rainfall_value`: ค่าปริมาณน้ำฝนที่ตรวจวัดได้ (มิลลิเมตร)
*   `tele_station_id`: รหัสประจำตัวของสถานีตรวจวัดทางไกล (Tele Station ID)
*   `tele_station_oldcode`: รหัสเก่าของสถานีตรวจวัดทางไกล
*   `measure_datetime`: วันที่ที่ทำการตรวจวัด (โดยไม่มีข้อมูลเวลา)

#Check Missing Data ตรวจสอบค่าว่าง

In [14]:
import numpy as np

api_dataframes = {
    'API07': API07,
    'API08': API08,
    'API09': API09,
    'API10': API10,
    'API11': API11,
    'API12': API12
}

for name, df in api_dataframes.items():
    print(f"\nMissing values in {name}:")
    print(df.isnull().sum())
    # Also check for empty strings if any numerical columns might have them as missing indicators
    # For relevant columns, check if they contain 'NaN' or other common missing value representations
    for col in df.columns:
        if df[col].dtype == 'object': # Check object type columns for empty strings
            empty_string_count = (df[col] == '').sum()
            if empty_string_count > 0:
                print(f"  Empty strings in '{col}': {empty_string_count}")


print("\nSummary of missing values across all API DataFrames completed.")


Missing values in API07:
rainfall_datetime           0
rainfall_datetime_calc      0
rainfall_value              0
tele_station_id             0
tele_station_oldcode      428
measure_datetime            0
dtype: int64

Missing values in API08:
rainfall_datetime          1268
rainfall_datetime_calc     1268
rainfall_value             1271
tele_station_id            2536
tele_station_oldcode       1832
measure_datetime              0
station_code              27915
value                     27915
dtype: int64

Missing values in API09:
rainfall_datetime          2546
rainfall_datetime_calc     2546
rainfall_value             2546
tele_station_id            5092
tele_station_oldcode       3777
measure_datetime              0
station_code              27796
value                     27796
dtype: int64

Missing values in API10:
rainfall_datetime            0
rainfall_datetime_calc       0
rainfall_value               1
tele_station_id              0
tele_station_oldcode      2183
measure_da

#2. กรองข้อมูลเฉพาะสถานีที่อยู่ใน Stations_model_HII
ขั้นตอนนี้เราจะตัดข้อมูลใน hii_daily_rain ที่ไม่มีรายชื่ออยู่ใน Stations_model_HII ออก

In [17]:
for i in Stations_model_HII.columns:
    print(i)

Node
Lat
Long


In [18]:
for i in hii_daily_rain:
    print(i)

station_code
measure_datetime
data
station_name
latitude
longitude
quality_flag
basin
sub_basin
tambon
amphoe
province
data_type


In [22]:
set(Stations_model_HII['Node'])

{'ACRU',
 'BBHN',
 'BBUA',
 'BCNG',
 'BDAR',
 'BDCP',
 'BDGN',
 'BDLH',
 'BDMG',
 'BGRT',
 'BHMN',
 'BHRA',
 'BHUN',
 'BJIG',
 'BKDN',
 'BKHL',
 'BKHN',
 'BKK001',
 'BKK008',
 'BKK009',
 'BKK015',
 'BKK017',
 'BKK018',
 'BKK019',
 'BKUG',
 'BKWN',
 'BLAT',
 'BLKO',
 'BLUG',
 'BMDG',
 'BMNK',
 'BNAN',
 'BNGR',
 'BNHG',
 'BNHI',
 'BNHO',
 'BNKN',
 'BNKP',
 'BNLG',
 'BNMK',
 'BNPI',
 'BNPU',
 'BNTK',
 'BOKA',
 'BPIA',
 'BPK004',
 'BPLA',
 'BPPS',
 'BPTG',
 'BRKM',
 'BSBA',
 'BSDE',
 'BSKJ',
 'BSMP',
 'BSPN',
 'BSUM',
 'BTHO',
 'BWKG',
 'BYNU',
 'CATK',
 'CBUR',
 'CCSS',
 'CEHM',
 'CGDO',
 'CGKM',
 'CGMN',
 'CGSN',
 'CHHA',
 'CHI005',
 'CHI010',
 'CHI011',
 'CHI012',
 'CHI013',
 'CHM001',
 'CHM002',
 'CHM003',
 'CHM004',
 'CHN004',
 'CHNA',
 'CHPA',
 'CHR003',
 'CHR004',
 'CHR005',
 'CHUN',
 'CLPK',
 'CMSG',
 'CPKC',
 'CPY003',
 'CPY004',
 'CPY009',
 'CPY010',
 'CPY012',
 'CPY013',
 'CPY015',
 'CPY017',
 'CSMO',
 'CTKN',
 'DCCT',
 'DGSL',
 'DITO',
 'DIV001',
 'DIV002',
 'DIV003',
 'DIV004'

In [16]:
set(hii_daily_rain['station_code'])

{'CHNA',
 'CGMN',
 'MOU305',
 'PASG',
 'MYA003',
 'FOP034',
 'HDA004',
 'NOKK',
 'ONE022',
 'WGHG',
 'MOU493',
 'ATG042',
 'BKNH',
 'YOM013',
 'MOU486',
 'MOU011',
 'NKO004',
 'VLGE48',
 'VLGE19',
 'ONE067',
 'CKLG',
 'MOU111',
 'STH031',
 'NBON',
 'NKO002',
 'ONE034',
 'SENM',
 'NEAK',
 'MSRM',
 'KHHM',
 'SKN003',
 'PNRN',
 'KAHN',
 'PCH001',
 'KON004',
 'BNPI',
 'ONE036',
 'BUKT',
 'FOP024',
 'MECN',
 'ONE091',
 'TCP010',
 'MOU128',
 'NKO003',
 'KUBR',
 'ATG111',
 'PRGB',
 'TCP003',
 'MYA002',
 'RGKG',
 'ONE021',
 'MOU034',
 'KKCN',
 'MOU499',
 'ONE094',
 'MOU288',
 'STH003',
 'STU001',
 'ONE017',
 'ONE044',
 'ONE030',
 'CPY002',
 'KKPA',
 'ONE110',
 'PYO004',
 'OMKI',
 'CHM001',
 'MOU024',
 'VLGE43',
 'LNSK',
 'MOU472',
 'TB0006',
 'MOU423',
 'DKLG',
 'PRST',
 'MEPK',
 'AIT009',
 'TKFA',
 'MOU175',
 'NKOK',
 'PANC',
 'BSDE',
 'ONE088',
 'NONA',
 'WEI072',
 'PAS001',
 'BKDN',
 'MOU130',
 'ONE040',
 'THA004',
 'MOU351',
 'PMPA',
 'MOU141',
 'NAN009',
 'YOM010',
 'MOU107',
 'BKK008',
 

In [26]:
Stations_model_HII.shape

(385, 3)

In [21]:
hii_daily_rain.shape

(3768463, 13)

In [27]:
import pandas as pd
import numpy as np

# แปลงวันที่ให้เป็น datetime object เพื่อความชัวร์
hii_daily_rain['measure_datetime'] = pd.to_datetime(hii_daily_rain['measure_datetime'])

# 1. ดึงรายชื่อสถานีเป้าหมาย (Nodes)
target_stations = Stations_model_HII['Node'].unique()

# นับจำนวนก่อนกรอง
count_before = len(hii_daily_rain)
stations_before = hii_daily_rain['station_code'].nunique()

# กรองข้อมูล
df_filtered = hii_daily_rain[hii_daily_rain['station_code'].isin(target_stations)].copy()

# นับจำนวนหลังกรอง
count_after = len(df_filtered)
stations_after = df_filtered['station_code'].nunique()

print(f"--- ผลการกรองสถานี ---")
print(f"จำนวน Record ก่อนกรอง: {count_before} (จาก {stations_before} สถานี)")
print(f"จำนวน Record หลังกรอง: {count_after} (เหลือ {stations_after} สถานี)")

--- ผลการกรองสถานี ---
จำนวน Record ก่อนกรอง: 3768463 (จาก 1387 สถานี)
จำนวน Record หลังกรอง: 1662282 (เหลือ 385 สถานี)


#3. ระบุและจัดการข้อมูลสูญหาย (Missing Data)
การหา "Record ที่หายไป" (Missing Records) ในข้อมูล Time Series ต้องสร้างช่วงเวลา (Date Range) ที่สมบูรณ์ขึ้นมาก่อน เพื่อดูว่าวันไหนหายไป

In [28]:
# สร้างช่วงเวลาตั้งแต่วันแรกถึงวันสุดท้ายที่มีข้อมูล
min_date = df_filtered['measure_datetime'].min()
max_date = df_filtered['measure_datetime'].max()
full_date_range = pd.date_range(start=min_date, end=max_date, freq='D')

# สร้าง MultiIndex ของ (ทุกสถานี x ทุกวัน)
all_stations = df_filtered['station_code'].unique()
mux = pd.MultiIndex.from_product([all_stations, full_date_range], names=['station_code', 'measure_datetime'])

# Reindex ข้อมูลเดิมเข้ากับ Index ใหม่ (วันที่ที่ไม่มีข้อมูลจะเป็น NaN)
df_full = df_filtered.set_index(['station_code', 'measure_datetime']).reindex(mux).reset_index()

# คำนวณจำนวนและสัดส่วนข้อมูลที่หายไป (นับจากคอลัมน์ data ที่เป็น NaN)
missing_stats = df_full.groupby('station_code')['data'].apply(lambda x: x.isnull().sum())
total_records = len(full_date_range)
missing_percent = (missing_stats / total_records) * 100

# แสดงตัวอย่าง
print("--- สถิติข้อมูลสูญหายรายสถานี (บางส่วน) ---")
print(missing_percent.head())

--- สถิติข้อมูลสูญหายรายสถานี (บางส่วน) ---
station_code
ACRU    18.405428
BBHN    20.695505
BBUA    16.200170
BCNG    15.754877
BDAR    21.055980
Name: data, dtype: float64


#4. คัดสถานีที่มีข้อมูลสูญหาย $\ge 20\%$ ออก

In [29]:
# หา list รายชื่อสถานีที่ข้อมูลหายมากกว่าหรือเท่ากับ 20%
stations_to_drop = missing_percent[missing_percent >= 20].index
stations_to_keep = missing_percent[missing_percent < 20].index

# แสดงผลการคัดออก
print(f"--- การคัดเลือกสถานี ---")
print(f"จำนวนสถานีที่ถูกคัดออก: {len(stations_to_drop)}")
print(f"รายชื่อสถานีที่ถูกคัดออก: {list(stations_to_drop)}")
print(f"จำนวนสถานีที่เหลืออยู่: {len(stations_to_keep)}")

# กรอง DataFrame ให้เหลือเฉพาะสถานีที่ผ่านเกณฑ์
df_clean = df_full[df_full['station_code'].isin(stations_to_keep)].copy()

--- การคัดเลือกสถานี ---
จำนวนสถานีที่ถูกคัดออก: 223
รายชื่อสถานีที่ถูกคัดออก: ['BBHN', 'BDAR', 'BHMN', 'BHRA', 'BKHL', 'BKHN', 'BKK001', 'BKK008', 'BKK009', 'BKK015', 'BKK017', 'BKK018', 'BKK019', 'BKUG', 'BKWN', 'BLAT', 'BLKO', 'BNAN', 'BNHG', 'BNHI', 'BNMK', 'BNPI', 'BNTK', 'BPIA', 'BPK004', 'BPLA', 'BPPS', 'BPTG', 'BRKM', 'BSKJ', 'BSMP', 'BSUM', 'BYNU', 'CEHM', 'CGKM', 'CGMN', 'CGSN', 'CHI005', 'CHI010', 'CHI011', 'CHI012', 'CHI013', 'CHM001', 'CHM003', 'CHM004', 'CHN004', 'CHR003', 'CHR004', 'CHR005', 'CHUN', 'CMSG', 'CPKC', 'CPY003', 'CPY009', 'CPY012', 'CPY013', 'CPY015', 'CPY017', 'CTKN', 'DGSL', 'DITO', 'DIV001', 'DIV002', 'DIV003', 'DIV004', 'DIV005', 'DKLG', 'DMKH', 'HCTT', 'HKTG', 'HNKA', 'HYOD', 'HYRT', 'JBLG', 'KANG', 'KAOP', 'KGKG', 'KHHM', 'KHPB', 'KKHA', 'KKTO', 'KMSI', 'KNYM', 'KOKN', 'KON002', 'KON004', 'KON005', 'KOSK', 'KRN002', 'KTKB', 'KWSI', 'LBI002', 'LNRG', 'MCRM', 'MDTG', 'MECN', 'MEWG', 'MRMT', 'MUN011', 'MUN014', 'NAKU', 'NAN010', 'NAN011', 'NAN012', 'NAN01

#5. เติมข้อมูลที่หายไปด้วยค่า Median (Day of Year)
เราจะใช้ค่ามัธยฐานของ "วันที่ของปี" (เช่น ค่าเฉลี่ยของทุกวันที่ 1 มกราคม ของสถานีนั้นๆ) มาเติมลงในช่องว่าง

In [30]:
# สร้างคอลัมน์ Day of Year (1-366)
df_clean['doy'] = df_clean['measure_datetime'].dt.dayofyear

# คำนวณ Median แยกตาม สถานี และ Day of Year
# transform จะช่วยกระจายค่ากลับไปให้เท่ากับจำนวนแถวเดิม เพื่อให้ง่ายต่อการ fillna
median_values = df_clean.groupby(['station_code', 'doy'])['data'].transform('median')

# เติมข้อมูล: ถ้า data ว่าง ให้ใช้ค่าจาก median_values, ถ้ามีค่าอยู่แล้วให้ใช้ค่าเดิม
df_clean['data'] = df_clean['data'].fillna(median_values)

# กรณีที่ Median ก็ยังเป็น NaN (เช่น วันที่ 29 ก.พ. ไม่เคยมีฝนเลยตลอดประวัติ) อาจเติมด้วย 0 หรือ Forward Fill
df_clean['data'] = df_clean['data'].fillna(0)

# ตรวจสอบว่าไม่มี record หายอีก
remaining_nulls = df_clean['data'].isnull().sum()
print(f"--- ตรวจสอบการเติมข้อมูล ---")
print(f"จำนวนค่าว่างที่เหลืออยู่: {remaining_nulls}")

--- ตรวจสอบการเติมข้อมูล ---
จำนวนค่าว่างที่เหลืออยู่: 0


#6.อภิปรายเหตุผล: ทำไมจึงใช้ Median แทน Mean?
ในการจัดการข้อมูลทางอุทกวิทยา (Hydrology) หรือปริมาณน้ำฝน เหตุผลหลักที่นิยมใช้ Median (ค่ามัธยฐาน) มากกว่า Mean (ค่าเฉลี่ย) มีดังนี้:

* การกระจายตัวของข้อมูล (Skewed Distribution): ข้อมูลปริมาณน้ำฝนมักมีการแจกแจงแบบเบ้ขวา (Right-skewed) คือมีวันที่ฝนไม่ตก (0 mm) จำนวนมาก และมีวันที่ฝนตกหนักมากๆ (Extreme values) จำนวนน้อย

* ค่า Mean จะถูกดึงให้สูงขึ้นอย่างมากโดยค่าฝนตกหนักเพียงไม่กี่วัน ทำให้ค่าที่ได้ไม่สะท้อน "เหตุการณ์ปกติ"

* ค่า Median จะอยู่ตรงกลางของข้อมูลจริงๆ และได้รับผลกระทบจากค่าสุดโต่ง (Outliers) น้อยกว่า จึงเป็นตัวแทนของปริมาณฝนในวันที่ "ปกติ" ได้ดีกว่า

* ในกรณีที่มีค่าว่างไม่เยอะมาก จึงใช้คำที่พบบ่อยมากที่สุดในการจัดการค่าว่าง

* เพราะถ้าข้อมูลวันนั้นหายไป เราจะใช้ "ค่ากลางที่สถานีนั้นๆ มักจะวัดได้ในวันเดียวกัน" มาเสียบแทน ซึ่งเป็นวิธีที่น่าเชื่อถือกว่าการเติม 0 ทันที

#7.คำนวณผลรวมรายเดือนและจัดรูปแบบ Output

In [43]:
import pandas as pd

# Create a list containing the individual monthly DataFrames
api_dataframes_2025 = [API07, API08, API09, API10, API11, API12]

# Concatenate all DataFrames into a single DataFrame named API_2025
API_2025 = pd.concat(api_dataframes_2025, ignore_index=True)

print("Successfully concatenated all monthly API DataFrames into API_2025.")
print("First 5 rows of API_2025:")
print(API_2025.head())
print("Last 5 rows of API_2025:")
print(API_2025.tail())
print(f"Shape of API_2025: {API_2025.shape}")

Successfully concatenated all monthly API DataFrames into API_2025.
First 5 rows of API_2025:
           rainfall_datetime     rainfall_datetime_calc  rainfall_value  \
0  2025-07-02 00:00:00+07:00  2025-07-02T07:00:00+07:00             0.0   
1  2025-07-02 00:00:00+07:00  2025-07-02T07:00:00+07:00             0.0   
2  2025-07-02 00:00:00+07:00  2025-07-02T07:00:00+07:00            34.0   
3  2025-07-02 00:00:00+07:00  2025-07-02T07:00:00+07:00             0.2   
4  2025-07-02 00:00:00+07:00  2025-07-02T07:00:00+07:00            38.2   

   tele_station_id tele_station_oldcode measure_datetime station_code  value  
0            589.0                 MWOG       2025-07-01          NaN    NaN  
1            564.0                 MSRM       2025-07-01          NaN    NaN  
2              4.0               CPY015       2025-07-01          NaN    NaN  
3            867.0                 KMSI       2025-07-01          NaN    NaN  
4             54.0               CPY016       2025-07-01    

รวมข้อมูลรายเดือนทั้งหมดเข้าด้วยกันเป็นตารางรายปี (หรือรายครึ่งปี)

รายละเอียดการทำงานทีละส่วน
* api_dataframes_2025 = [...] : เป็นการรวบรวมตัวแปร DataFrame ของแต่ละเดือน (กรกฎาคม - ธันวาคม) มาใส่ไว้ใน List เพื่อเตรียมส่งต่อให้คำสั่งรวมข้อมูล

* pd.concat(..., ignore_index=True) : หัวใจหลัก: นำข้อมูลจากเดือน 07 ไปจนถึง 12 มาวางซ้อนต่อท้ายกัน

* ignore_index=True: สำคัญมาก เพราะจะทำให้เลข Index ของแถวเรียงใหม่ตั้งแต่ 0 ไปจนถึงแถวสุดท้ายของปี 2025 ป้องกันไม่ให้เลข Index ของแต่ละเดือนมาตีกันเอง

* API_2025.shape : แสดงขนาดของตารางใหม่ (จำนวนแถว, จำนวนคอลัมน์) เพื่อให้คุณเช็คว่าจำนวนข้อมูลทั้งหมดมีปริมาณเท่าไหร่หลังจากรวมเสร็จแล้ว

**ผลลัพธ์ที่ได้และความสำคัญ**

เมื่อรันโค้ดนี้เสร็จ คุณจะได้ตาราง API_2025 ที่เป็น "Master File" ของปีนั้น ซึ่งมีข้อดีคือ:

* วิเคราะห์ข้ามเดือนได้: เช่น การเปรียบเทียบปริมาณฝนระหว่างเดือนสิงหาคมกับกันยายน หรือการดูแนวโน้ม (Trend) ตลอดทั้ง 6 เดือน

* จัดการข้อมูลครั้งเดียวจบ: หากคุณต้องการลบค่าว่าง (Missing values) หรือเปลี่ยนรูปแบบวันที่ คุณสามารถทำที่ตาราง API_2025 เพียงที่เดียว แทนที่จะต้องไปทำแยก 6 ตาราง

## Clean and Standardize API Data

### Subtask:
Select the `tele_station_oldcode`, `measure_datetime`, and `rainfall_value` columns from the combined `API_2025` DataFrame. Rename `tele_station_oldcode` to `station_code` and `rainfall_value` to `data`. Convert `measure_datetime` to datetime objects and drop any rows where `station_code` is missing.


In [45]:
API_2025_cleaned = API_2025[['tele_station_oldcode', 'measure_datetime', 'rainfall_value']].copy()
API_2025_cleaned = API_2025_cleaned.rename(columns={'tele_station_oldcode': 'station_code', 'rainfall_value': 'data'})
API_2025_cleaned['measure_datetime'] = pd.to_datetime(API_2025_cleaned['measure_datetime'], format='mixed')
API_2025_cleaned.dropna(subset=['station_code'], inplace=True)

print("API_2025 DataFrame has been cleaned and standardized.")
print("First 5 rows of the cleaned DataFrame:")
print(API_2025_cleaned.head())
print("Shape of the cleaned DataFrame: " + str(API_2025_cleaned.shape))

API_2025 DataFrame has been cleaned and standardized.
First 5 rows of the cleaned DataFrame:
  station_code measure_datetime  data
0         MWOG       2025-07-01   0.0
1         MSRM       2025-07-01   0.0
2       CPY015       2025-07-01  34.0
3         KMSI       2025-07-01   0.2
4       CPY016       2025-07-01  38.2
Shape of the cleaned DataFrame: (168021, 3)


**คัดเลือกและปรับแต่งข้อมูล (Data Selection & Cleaning)**

เพื่อเตรียมความพร้อมสำหรับขั้นตอนการวิเคราะห์จริงครับ เป็นการเปลี่ยนจากตารางดิบ (Raw Data) ให้กลายเป็นตารางมาตรฐานที่ใช้งานง่ายขึ้น

* สิ่งที่โค้ดนี้ทำทีละขั้นตอน
1. คัดเลือกเฉพาะคอลัมน์ที่จำเป็น (.copy())
* เลือกมาเพียง 3
คอลัมน์คือ รหัสสถานี (tele_station_oldcode), วันเวลาที่วัด (measure_datetime), และปริมาณฝน (rainfall_value)
* การใช้ .copy() เพื่อสร้างตารางใหม่แยกออกมาอย่างเด็ดขาด ป้องกันการเกิด Warning เวลาเราแก้ไขข้อมูลในขั้นตอนถัดไป
2. เปลี่ยนชื่อคอลัมน์ (.rename)
* เปลี่ยนชื่อให้สั้นและสื่อความหมายง่ายขึ้น
* tele_station_oldcode $\rightarrow$ station_code
* rainfall_value $\rightarrow$ data
3. จัดการรูปแบบวันที่ (pd.to_datetime)
* แปลงคอลัมน์ measure_datetime จากตัวหนังสือ (String) ให้เป็นข้อมูล วันที่และเวลา (Datetime object)
* การใส่ format='mixed' ช่วยให้ Pandas ฉลาดพอที่จะจัดการได้แม้ว่ารูปแบบวันที่ในแต่ละแถวจะเขียนมาไม่เหมือนกัน (เช่น บางแถวเป็นวัน/เดือน/ปี บางแถวเป็น ปี-เดือน-วัน)
4. ลบข้อมูลที่ระบุสถานีไม่ได้ (.dropna)
* ลบทุกแถวที่ไม่มีรหัสสถานี (station_code เป็นค่าว่าง) ทิ้งไปทันที เพราะถ้าไม่รู้ว่าฝนตกที่ไหน ข้อมูลนั้นก็เอาไปวิเคราะห์ต่อไม่ได้

## Filter API Stations กรองข้อมูล

### Subtask:
Filter the cleaned API_2025 DataFrame to include only those stations whose `station_code` (originally `tele_station_oldcode`) is present in the `Node` column of the `Stations_model_HII` DataFrame.


In [46]:
target_api_stations = Stations_model_HII['Node'].unique()

API_2025_filtered = API_2025_cleaned[API_2025_cleaned['station_code'].isin(target_api_stations)].copy()

print("API_2025_cleaned DataFrame filtered to include only stations present in Stations_model_HII.")
print("First 5 rows of API_2025_filtered:")
print(API_2025_filtered.head())
print("Shape of API_2025_filtered: " + str(API_2025_filtered.shape))

API_2025_cleaned DataFrame filtered to include only stations present in Stations_model_HII.
First 5 rows of API_2025_filtered:
   station_code measure_datetime  data
2        CPY015       2025-07-01  34.0
3          KMSI       2025-07-01   0.2
7          CSMO       2025-07-01   0.0
8          NPCT       2025-07-01   0.0
10       KON005       2025-07-01  44.0
Shape of API_2025_filtered: (66275, 3)


**กรองข้อมูล (Filtering)**
เพื่อเลือกเอาเฉพาะสถานีวัดน้ำฝนที่คุณสนใจจริงๆ โดยอ้างอิงจากรายชื่อสถานีที่มีอยู่ในอีกตารางหนึ่ง

1. การสร้าง "รายชื่อเป้าหมาย"
Stations_model_HII['Node'].unique():

* เป็นการดึงค่าจากคอลัมน์ Node (ซึ่งน่าจะเป็นรหัสสถานีในโมเดล HII)

* ใช้ .unique() เพื่อเอาเฉพาะรายชื่อที่ไม่ซ้ำกัน

* ผลลัพธ์ที่ได้คือ target_api_stations ซึ่งเปรียบเสมือน "White List" หรือรายชื่อสถานีที่เราต้องการเก็บไว้เท่านั้น

2. การคัดกรองข้อมูล
.isin(target_api_stations):

* คำสั่งนี้จะวิ่งไปเช็คในตาราง API_2025_cleaned (ตารางรวมปี 2025 ที่คุณคลีนมาแล้ว)

* มันจะตรวจสอบว่า station_code ในแต่ละแถว "มีชื่ออยู่ในรายชื่อเป้าหมายหรือไม่"

* API_2025_cleaned[...] : ทำการเลือก (Filter) เอาเฉพาะแถวที่เป็นจริง (คือมีรหัสสถานีตรงกับกลุ่มเป้าหมาย) ออกมา

3. การบันทึกผล
* .copy(): เป็นการสร้าง DataFrame ใหม่ที่ชื่อว่า API_2025_filtered อย่างสมบูรณ์ เพื่อไม่ให้ไปรบกวนข้อมูลเดิมและป้องกัน Warning ในการทำงานขั้นต่อไป


## Handle Missing Daily Records (2025)

### Subtask:
Create a complete date range for the entire year 2025. Then, reindex the filtered API_2025 DataFrame with a MultiIndex of all filtered stations and this full date range to explicitly show any missing daily records.


In [47]:
import pandas as pd

# 1. Extract the minimum and maximum dates from the measure_datetime column
min_date_api = API_2025_filtered['measure_datetime'].min()
max_date_api = API_2025_filtered['measure_datetime'].max()

# 2. Generate a complete date range
full_date_range_api = pd.date_range(start=min_date_api, end=max_date_api, freq='D')

# 3. Identify all unique station_code values
all_stations_api = API_2025_filtered['station_code'].unique()

# 4. Create a MultiIndex of (all stations x every day)
mux_api = pd.MultiIndex.from_product([all_stations_api, full_date_range_api], names=['station_code', 'measure_datetime'])

# 5. Set the station_code and measure_datetime columns as the index and reindex it with the newly created MultiIndex
# 6. Reset the index of the reindexed DataFrame to convert the MultiIndex back into columns.
# 7. Store the resulting DataFrame in a new variable, API_2025_full
API_2025_full = API_2025_filtered.set_index(['station_code', 'measure_datetime']).reindex(mux_api).reset_index()

print("Successfully created API_2025_full with a complete date range for each station.")
print("First 5 rows of API_2025_full:")
print(API_2025_full.head())
print("Last 5 rows of API_2025_full:")
print(API_2025_full.tail())
print(f"Shape of API_2025_full: {API_2025_full.shape}")

Successfully created API_2025_full with a complete date range for each station.
First 5 rows of API_2025_full:
  station_code measure_datetime  data
0       CPY015       2025-07-01  34.0
1       CPY015       2025-07-02   1.6
2       CPY015       2025-07-03   0.2
3       CPY015       2025-07-04   0.0
4       CPY015       2025-07-05   1.2
Last 5 rows of API_2025_full:
      station_code measure_datetime  data
70651         CHNA       2025-12-27   8.6
70652         CHNA       2025-12-28   1.2
70653         CHNA       2025-12-29   0.0
70654         CHNA       2025-12-30   0.0
70655         CHNA       2025-12-31   0.0
Shape of API_2025_full: (70656, 3)


Data Interpolation (Time-Series)" หรือการขยายตารางข้อมูลเพื่อให้ "ทุกสถานีมีข้อมูลครบทุกวัน

**รายละเอียดการทำงานทีละขั้นตอน**
1. หาช่วงเวลา (Min/Max Date): หาจุดเริ่มต้นและจุดสิ้นสุดของข้อมูลทั้งหมดในเซตนี้
2. สร้างปฏิทินอ้างอิง (pd.date_range): สร้างรายการวันที่เรียงกันแบบ "รายวัน" (freq='D') ตั้งแต่วันแรกจนถึงวันสุดท้ายแบบไม่ข้ามแม้แต่วันเดียว
3. หาจำนวนสถานีทั้งหมด: ดึงรายชื่อสถานีที่ไม่ซ้ำกันออกมา
4. สร้างตารางเปล่าแบบ "คูณไขว้" (pd.MultiIndex):
* ขั้นตอนนี้คือหัวใจสำคัญครับ มันจะเอา (รายชื่อทุกสถานี) × (ทุกวันที่ในปฏิทิน)
* สมมติมี 10 สถานี และมี 180 วัน คุณจะได้ตารางที่มี $10 \times 180 = 1,800$ แถวแน่นอน
5. การ reindex
* นำตารางจริงของคุณไปวางทับบนตารางเปล่าที่เพิ่งสร้าง
* แถวไหนที่มีข้อมูลอยู่แล้ว (มีค่าฝน) ข้อมูลก็จะยังอยู่แถวไหนที่ในตารางจริงไม่มี (เช่น วันที่ 5 สถานี A หายไป): Pandas จะสร้างแถวของวันที่ 5 สถานี A ขึ้นมาให้ใหม่โดยอัตโนมัติ แต่ค่าปริมาณฝนจะกลายเป็น NaN (ค่าว่าง)

## Drop Stations with Excessive Missing Data (2025)




In [49]:
# 1. Calculate the number of missing `data` values for each `station_code`
missing_data_count_api = API_2025_full.groupby('station_code')['data'].apply(lambda x: x.isnull().sum())

# 2. Determine the total number of days in the `full_date_range_api`
total_days_api = len(full_date_range_api)

# 3. Calculate the percentage of missing `data` for each station
missing_percent_api = (missing_data_count_api / total_days_api) * 100

# 4. Identify stations to drop (>= 20% missing)
stations_to_drop_api = missing_percent_api[missing_percent_api >= 20].index.tolist()

# 5. Identify stations to keep (< 20% missing)
stations_to_keep_api = missing_percent_api[missing_percent_api < 20].index.tolist()

# 6. Print the results
print(f"--- API 2025 Station Filtering ---")
print(f"Total stations: {len(all_stations_api)}")
print(f"Number of stations to be dropped (>= 20% missing): {len(stations_to_drop_api)}")
print(f"Stations to be dropped: {stations_to_drop_api}")
print(f"Number of stations to be kept (< 20% missing): {len(stations_to_keep_api)}")
print(f"Stations to be kept: {stations_to_keep_api}")

# 7. Filter the DataFrame to keep only the desired stations
API_2025_clean = API_2025_full[API_2025_full['station_code'].isin(stations_to_keep_api)].copy()

print("\nFirst 5 rows of API_2025_clean after filtering:")
print(API_2025_clean.head())
print(f"Shape of API_2025_clean: {API_2025_clean.shape}")

--- API 2025 Station Filtering ---
Total stations: 384
Number of stations to be dropped (>= 20% missing): 45
Stations to be dropped: ['BKHN', 'BKK017', 'BLKO', 'BNGR', 'BNLG', 'BPTG', 'BSKJ', 'BSPN', 'CEHM', 'CGKM', 'CHI010', 'CHI013', 'CHNA', 'CPY010', 'CTKN', 'DCCT', 'DGSL', 'KKTO', 'KON005', 'LMSK', 'MEWG', 'MMMO', 'MNRM', 'NAPK', 'NASW', 'NEAK', 'NGCG', 'NGPI', 'NKSG', 'NNPT', 'NPI002', 'PAS008', 'PMPN', 'PRN001', 'PSLT', 'SGKW', 'SKE001', 'SKM004', 'SLA002', 'TPI007', 'TUKW', 'VLGE33', 'WAN005', 'YCUM', 'YOM001']
Number of stations to be kept (< 20% missing): 339
Stations to be kept: ['ACRU', 'BBHN', 'BBUA', 'BCNG', 'BDAR', 'BDCP', 'BDGN', 'BDLH', 'BDMG', 'BGRT', 'BHMN', 'BHRA', 'BHUN', 'BJIG', 'BKDN', 'BKHL', 'BKK001', 'BKK008', 'BKK015', 'BKK018', 'BKK019', 'BKUG', 'BKWN', 'BLAT', 'BLUG', 'BMDG', 'BMNK', 'BNAN', 'BNHG', 'BNHI', 'BNHO', 'BNKN', 'BNKP', 'BNMK', 'BNPI', 'BNPU', 'BNTK', 'BOKA', 'BPIA', 'BPK004', 'BPLA', 'BPPS', 'BRKM', 'BSBA', 'BSDE', 'BSMP', 'BSUM', 'BTHO', 'BWKG',

คัดกรองสถานีวัดน้ำฝนตามเกณฑ์คุณภาพข้อมูล

1. การคำนวณหา "รูโหว่" ของข้อมูล
* missing_data_count_api: โค้ดจะเข้าไปนับว่าในแต่ละสถานี มีจำนวนกี่วันที่ไม่มีการบันทึกค่าปริมาณฝน (ค่าเป็น NaN หลังจากที่เราทำ reindex ในขั้นตอนก่อนหน้า)

* missing_percent_api: นำจำนวนวันที่หายไปมาหารด้วยจำนวนวันทั้งหมดในรอบปีนั้น (total_days_api) แล้วคูณ 100 เพื่อให้ได้ เปอร์เซ็นต์ของข้อมูลที่หายไป (Missing Percentage)

2. การกำหนดเกณฑ์การคัดออก (Threshold)
โค้ดนี้ใช้เกณฑ์ความน่าเชื่อถือที่ 80% หมายความว่า:

* stations_to_drop_api: สถานีใดที่มีข้อมูลหายไปตั้งแต่ 20% ขึ้นไป จะถือว่ามีข้อมูลไม่ต่อเนื่องพอที่จะนำไปวิเคราะห์ผลรวมรายเดือนหรือรายปี จึงถูกเก็บชื่อไว้ในรายการ "คัดออก"

* stations_to_keep_api: สถานีที่มีข้อมูลเกือบครบ (หายไปน้อยกว่า 20%) จะถูกเก็บไว้ใช้งานต่อ

3. การกรองตารางขั้นสุดท้าย
* API_2025_clean: สร้างตารางใหม่โดยการเลือก (.isin) เอาเฉพาะสถานีที่ผ่านเกณฑ์คุณภาพข้อมูลเท่านั้น

* .copy(): เพื่อให้มั่นใจว่าข้อมูลชุดที่ "สะอาดแล้ว" นี้เป็นอิสระจากตารางเก่า ไม่มีการเชื่อมโยงที่อาจทำให้เกิด Error ในภายหลัง

## Impute Missing Rainfall Values (2025)




In [50]:
API_2025_clean['doy'] = API_2025_clean['measure_datetime'].dt.dayofyear

median_values_api = API_2025_clean.groupby(['station_code', 'doy'])['data'].transform('median')

API_2025_clean['data'] = API_2025_clean['data'].fillna(median_values_api)

API_2025_clean['data'] = API_2025_clean['data'].fillna(0)

remaining_nulls_api = API_2025_clean['data'].isnull().sum()
print(f"--- ตรวจสอบการเติมข้อมูล API 2025 ---")
print(f"จำนวนค่าว่างที่เหลืออยู่: {remaining_nulls_api}")

--- ตรวจสอบการเติมข้อมูล API 2025 ---
จำนวนค่าว่างที่เหลืออยู่: 0


เติมค่าว่าง (Imputation)

รายละเอียดการทำงานทีละส่วน
1. สร้างคอลัมน์ doy (Day of Year):

* API_2025_clean['measure_datetime'].dt.dayofyear จะแปลงวันที่ให้เป็นตัวเลขลำดับวันของปี (1-366)

* ตัวอย่าง: วันที่ 1 กุมภาพันธ์ จะถูกแปลงเป็นวันที่ 32 ของปี

2. คำนวณค่ามัธยฐานรายสถานีแยกตามวัน (transform('median')):

* โค้ดจะหาค่า Median (ค่ากลาง) ของปริมาณน้ำฝน โดยจับคู่ระหว่าง "รหัสสถานี" และ "ลำดับวันของปี"

* การใช้ transform จะช่วยให้ค่ากลางที่คำนวณได้ถูกกระจายกลับไปวางในตำแหน่งเดิมของทุกแถวที่มีสถานีและวันตรงกัน เพื่อเตรียมพร้อมสำหรับการเติมค่า

3. การเติมค่าว่างรอบที่ 1 (Fillna with Median):

* API_2025_clean['data'].fillna(median_values_api) จะนำค่ามัธยฐานที่หาได้มาเติมในช่องที่เป็นว่าง (NaN)

* ทำไมถึงทำแบบนี้? เพราะถ้าข้อมูลวันนั้นหายไป เราจะใช้ "ค่ากลางที่สถานีนั้นๆ มักจะวัดได้ในวันเดียวกัน" มาเสียบแทน ซึ่งเป็นวิธีที่น่าเชื่อถือกว่าการเติม 0 ทันที

4.การเติมค่าว่างรอบที่ 2 (Fillna with 0):

* หากรอบแรกยังเหลือค่าว่างอยู่ (เช่น สถานีนั้นไม่เคยมีข้อมูลของวันนั้นเลยจนหาค่ามัธยฐานไม่ได้) จะทำการเติมด้วย 0 เพื่อปิดรูโหว่ทั้งหมดให้สนิท

## Calculate Monthly Rainfall for 2025

### Subtask:
Extract the month and year from the `measure_datetime` column. Group the processed 2025 DataFrame by `station_code`, `year`, and `month`, and then sum the `data` (rainfall) to get the total monthly rainfall. Rename the resulting columns to `Node`, `rain`, `month`, and `year` to match the required output format, and order them as `Node`, `rain`, `month`, `year`.


In [51]:
API_2025_clean['month'] = API_2025_clean['measure_datetime'].dt.month
API_2025_clean['year'] = API_2025_clean['measure_datetime'].dt.year

API_2025_monthly_rainfall = API_2025_clean.groupby(['station_code', 'year', 'month'])['data'].sum().reset_index()

API_2025_monthly_rainfall = API_2025_monthly_rainfall.rename(columns={'station_code': 'Node', 'data': 'rain'})

API_2025_monthly_rainfall = API_2025_monthly_rainfall[['Node', 'rain', 'month', 'year']]

print("Monthly rainfall for 2025 calculated and formatted successfully.")
print("First 5 rows of API_2025_monthly_rainfall:")
print(API_2025_monthly_rainfall.head())
print(f"Shape of API_2025_monthly_rainfall: {API_2025_monthly_rainfall.shape}")

Monthly rainfall for 2025 calculated and formatted successfully.
First 5 rows of API_2025_monthly_rainfall:
   Node   rain  month  year
0  ACRU   83.0      7  2025
1  ACRU  164.8      8  2025
2  ACRU  248.6      9  2025
3  ACRU   48.4     10  2025
4  ACRU   52.2     11  2025
Shape of API_2025_monthly_rainfall: (2034, 4)


**ขั้นตอนสุดท้ายของการสรุปข้อมูล**

โดยเป็นการเปลี่ยนจากข้อมูลรายวัน (Daily) ให้กลายเป็น "ผลรวมปริมาณน้ำฝนรายเดือน (Monthly Total)

รายละเอียดการทำงานทีละขั้นตอน
1. การสกัดหน่วยเวลา

* API_2025_clean['month'] = ... และ API_2025_clean['year'] = ...

เป็นการสร้างคอลัมน์ใหม่เพื่อระบุว่าข้อมูลแต่ละแถวอยู่ในเดือนไหนและปีไหน เพื่อใช้เป็นเกณฑ์ในการรวมกลุ่ม (Grouping)

2. การรวมผลรวม (Aggregation):

* groupby(['station_code', 'year', 'month'])['data'].sum()

*  ระบบจะหยิบข้อมูลปริมาณฝน (data) ทั้งหมดที่อยู่ใน "สถานีเดียวกัน เดือนเดียวกัน และปีเดียวกัน" มาบวกกันทั้งหมด

* .reset_index() เปลี่ยนผลลัพธ์จากโครงสร้าง Groupby ให้กลับมาเป็นตาราง DataFrame ปกติ

3. การปรับรูปแบบตาราง (Formatting):

* เปลี่ยนชื่อคอลัมน์: เปลี่ยน station_code เป็น Node และ data เป็น rain เพื่อให้ตรงกับมาตรฐานของโมเดล HII หรือระบบที่คุณกำลังใช้งานอยู่

* จัดเรียงคอลัมน์: ใช้ [['Node', 'rain', 'month', 'year']] เพื่อจัดลำดับคอลัมน์ให้เป็นระเบียบตามที่ต้องการ

**ผลลัพธ์ที่ได้และความหมาย**

ตาราง API_2025_monthly_rainfall ที่ได้จะมีลักษณะดังนี้:

* แต่ละแถวคือ "ปริมาณฝนรวมทั้งเดือน" ของหนึ่งสถานี

* ข้อมูลจะดูสะอาดและนำไปพล็อตกราฟเปรียบเทียบรายเดือนได้ง่ายมาก

* Shape ของตารางจะเล็กลงอย่างมาก (เหลือจำนวนแถวเท่ากับ: จำนวนสถานี × จำนวนเดือน)

## Save 2025 Monthly Rainfall to Google Sheet




In [35]:
import sys
!{sys.executable} -m pip install gspread google-auth

In [38]:
import gspread
from google.colab import auth
from google.auth import default
import pandas as pd

In [39]:
# --- 1. ส่วนสำหรับเชื่อมต่อ Google Drive และ Google Sheets ---
print("กำลังเชื่อมต่อ Google Drive...")
drive.mount('/content/drive')

# นำเข้าไลบรารีสำหรับ Google Sheets และทำการยืนยันตัวตน
from google.colab import auth
auth.authenticate_user()

import gspread
from google.auth import default
from gspread_dataframe import set_with_dataframe

print("กำลังเชื่อมต่อ Google Sheets API...")
creds, _ = default()
gc = gspread.authorize(creds)
print("เชื่อมต่อสำเร็จ!")

กำลังเชื่อมต่อ Google Drive...
Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
กำลังเชื่อมต่อ Google Sheets API...
เชื่อมต่อสำเร็จ!


In [52]:
import gspread
from gspread_dataframe import set_with_dataframe

# Define the name of your new Google Sheet and Worksheet
sheet_name_2025 = '2025 Monthly Rainfall Data'
worksheet_name_2025 = 'Rainfall 2025'

try:
    # Try to open the spreadsheet by name
    spreadsheet_2025 = gc.open(sheet_name_2025)
    print(f"Successfully opened existing spreadsheet: {sheet_name_2025}")
except gspread.exceptions.SpreadsheetNotFound:
    # If not found, create a new one
    spreadsheet_2025 = gc.create(sheet_name_2025)
    print(f"Created new spreadsheet: {sheet_name_2025}")

# Select the worksheet, create if it doesn't exist
try:
    worksheet_2025 = spreadsheet_2025.worksheet(worksheet_name_2025)
    print(f"Successfully opened existing worksheet: {worksheet_name_2025}")
except gspread.exceptions.WorksheetNotFound:
    # Use the existing `gc` object to add a worksheet
    worksheet_2025 = spreadsheet_2025.add_worksheet(title=worksheet_name_2025, rows=API_2025_monthly_rainfall.shape[0]+1, cols=API_2025_monthly_rainfall.shape[1])
    print(f"Created new worksheet: {worksheet_name_2025}")

# Clear existing content and write DataFrame headers and data
set_with_dataframe(worksheet_2025, API_2025_monthly_rainfall, include_index=False)

print(f"Data successfully written to Google Sheet: {sheet_name_2025}, Worksheet: {worksheet_name_2025}")
print(f"You can access your sheet here: {spreadsheet_2025.url}")

Created new spreadsheet: 2025 Monthly Rainfall Data
Created new worksheet: Rainfall 2025
Data successfully written to Google Sheet: 2025 Monthly Rainfall Data, Worksheet: Rainfall 2025
You can access your sheet here: https://docs.google.com/spreadsheets/d/1mfsdgWVcli2MbMupTJw68XwFVqxk6W5oM-F4-iS9R8s


1. **การกรองข้อมูลให้เหลือเฉพาะสถานีที่อยู่ใน Stations_model_HII:**

จำนวน Record ก่อนกรอง: 3,768,463 แถว (จาก 1,387 สถานี)
จำนวน Record หลังกรอง: 1,662,282 แถว (เหลือ 385 สถานี)

2. **การระบุข้อมูลสูญหาย (record ที่หายไป) ของแต่ละสถานี**

* หลังจากสร้างช่วงเวลาเต็ม (full date range) ข้อมูลถูก reindex เพื่อให้เห็นค่าว่าง (NaN) สำหรับวันที่ไม่มีข้อมูล.
* ตัวอย่างสัดส่วนข้อมูลที่สูญหาย (Missing Percentage) สำหรับบางสถานี:
ACRU: 18.41%
BBHN: 20.70%
BBUA: 16.20%
BCNG: 15.75%
BDAR: 21.06%


3. **การคัดสถานีที่มีข้อมูลสูญหายมากกว่าหรือเท่ากับ 20% ออก**

* จำนวนสถานีที่ถูกคัดออก: 223 สถานี

* จำนวนสถานีที่เหลืออยู่: 162 สถานี


4. **การเติมข้อมูลที่หายไปด้วยค่า Median ตาม Day of year จาก hii_daily_rain**

* มีการสร้างคอลัมน์ doy (Day of Year) เพื่อระบุวันลำดับที่ของปี (1-366).
* คำนวณค่า Median ของปริมาณน้ำฝนแยกตาม station_code และ doy.
* เติมค่าว่างในคอลัมน์ data ด้วยค่า Median ที่คำนวณได้.
* ในกรณีที่ยังคงมีค่าว่างเหลืออยู่ (เช่น วันที่ไม่เคยมีข้อมูลเลย), ได้มีการเติมค่า 0 แทน.
* การตรวจสอบ: หลังจากเติมข้อมูลแล้ว จำนวนค่าว่างที่เหลืออยู่: 0.

5. **การคำนวณผลรวมปริมาณน้ำฝนรายเดือน**

* ขั้นตอนนี้ ได้ดำเนินการแล้ว โดยมีการสร้างคอลัมน์ year และ month จาก measure_datetime.
* จากนั้นข้อมูลถูก Grouped By ด้วย Node (station_code), year, และ month แล้วทำการ Sum ค่า data (ปริมาณน้ำฝน) เพื่อหาผลรวมรายเดือน.
* ผลลัพธ์คือ DataFrame ชื่อ output_df ที่มีคอลัมน์ Node, rain, month, และ year ครบถ้วน.